site/blog/unlisted/why-sql-for-infrastructure.mdx

```diff
@@ -0,0 +1,94 @@
+
+Until cloud resource management is as quick to execute as a webpage render where state changes can be done faster than a human can respond, this sort of hand-holding is necessary for live infrastructure with uptime guarantees. Declarative programming, at least for cloud infrastructure management, is a [leaky abstraction](https://en.wikipedia.org/wiki/Leaky_abstraction) over the imperative operations that need to be executed.
+
+The Terraform HCL files don't have the prior state encoded into them, nor does it depend on being stored within a git repository to provide that prior state, so how does Terraform get that initial state to operate on? Terraform maintains an [internal statefile](https://developer.hashicorp.com/terraform/language/state) to perform the diffing against. It has a mechanism to [refresh that state](https://developer.hashicorp.com/terraform/cli/commands/refresh), but being a JSON blob with poor typing, can [become corrupted, requiring manual intervention](https://faun.pub/cleaning-up-a-terraform-state-file-the-right-way-ab509f6e47f3).
+
+Why does Terraform need this JSON-based statefile when it has the much cleaner HCL format that you use as a developer? This is because HCL, as a language that provides affordances to the developer's own intuition on how to structure their representation of their infrastructure, does not have a [canonical form](https://en.wikipedia.org/wiki/Canonical_form) to make these comparisons with. Terraform must first pre-process the HCL into an internal data structure which it can then generate a [diff](https://en.wikipedia.org/wiki/Diff) that can then be used to determine the operations to perform. The JSON-based statefile is likely a very close representation of this internal data structure, and manual editing of it is highly discouraged because the details of it likely vary from release to release with an internal migration mechanism performed on upgrade of the Terraform CLI.
+
+Needing to learn Terraform's HCL language to make infrastructure changes has been cited as a [weakness](https://www.pulumi.com/why-pulumi/) that can be mitigated by wrapping it in the syntaxes more familiar to the user. This *is* a weakness, but we believe that the HCL-to-Statefile transformation being [surjective](https://en.wikipedia.org/wiki/Surjective_function) instead of [bijective](https://en.wikipedia.org/wiki/Bijection) is actually the bigger issue.
+
+<!-- TODO: This section could use actual HCL code, and ideally some way you can trigger a corruption of the statefile vs AWS's actual state -->
+
+## The Objective is Bi- hmmmm....
```

Comment:
I think there is a whole thread to pull on that could probably be its own section, or its own post about how we got to where we are in thinking that declarative is best for infrastructure. TF is basically a cloud-native Puppet. Puppet works in a world where data center resources are homogenous and static meaning most servers are alike, are programmed similarly and the amount of them doesn't change as often. However, in the cloud era there are 250+ services (or developer abstractions) only in AWS built on top of bare VMs which can be used on demand which makes a declarative approach much harder to define.

Als interesting that even then, an infra as data approach was already around https://www.techtarget.com/searchdatacenter/definition/configuration-management-database